vividus-analytics/src/main/java/org/vividus/analytics/GoogleAnalyticsFacade.java

```diff
@@ -0,0 +1,124 @@
+ * limitations under the License.
+ */
+
+package org.vividus.analytics;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.http.HttpEntity;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.methods.HttpEntityEnclosingRequestBase;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.message.BasicNameValuePair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.vividus.analytics.model.AnalyticsEvent;
+import org.vividus.http.HttpMethod;
+import org.vividus.http.client.IHttpClient;
+
+public class GoogleAnalyticsFacade
+{
+    private static final Pattern PROJECT_NAME_PATTERN = Pattern.compile("(.+)(-\\d.+)?(\\\\|/)?");
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GoogleAnalyticsFacade.class);
+
+    private URI analyticsUri;
+    private String trackingId;
+    private List<NameValuePair> defaultParameters;
+    private final IHttpClient httpClient;
+
+    public GoogleAnalyticsFacade(IHttpClient httpClient)
+    {
+        this.httpClient = httpClient;
+    }
+
+    public void init()
+    {
+        defaultParameters = new ArrayList<>();
+        defaultParameters.add(pairOf("v", "1"));
+        defaultParameters.add(pairOf("t", "event"));
+        defaultParameters.add(pairOf("tid", trackingId));
+        defaultParameters.add(pairOf("cid", cid()));
+    }
+
+    private static String cid()
+    {
+        String[] pathParts = System.getProperty("user.dir").split("\\\\|/");
+        String projectName = pathParts[pathParts.length - 1];
+        if ("scripts".equals(projectName))
+        {
+            projectName = PROJECT_NAME_PATTERN.matcher(pathParts[pathParts.length - 2]).replaceFirst("$1");
+        }
+        return UUID.nameUUIDFromBytes(DigestUtils.sha512Hex(projectName).getBytes(StandardCharsets.UTF_8)).toString();
+    }
+
+    public void postEvent(AnalyticsEvent analyticsEvent)
+    {
+        String params = Stream.concat(defaultParameters.stream(), convertToNameValuePairs(analyticsEvent))
+                              .map(p -> p.getName() + "=" + p.getValue())
+                              .collect(Collectors.joining("&"));
+        StringEntity entity = new StringEntity(params, StandardCharsets.UTF_8);
+        post(entity);
+    }
+
+    private Stream<BasicNameValuePair> convertToNameValuePairs(AnalyticsEvent analyticsEvent)
+    {
+        return analyticsEvent.getPayload()
+                             .entrySet()
+                             .stream()
+                             .map(e -> pairOf(e.getKey(), URLEncoder.encode(e.getValue(), StandardCharsets.UTF_8)));
```

Comment:
I recommend to use `org.apache.http.client.utils.URLEncodedUtils`

Reply:
Have you considered: `org.apache.http.client.utils.URLEncodedUtils#format(java.util.List<? extends org.apache.http.NameValuePair>, char, java.lang.String)` ?